ambry-replication/src/main/java/com.github.ambry.replication/ReplicaThread.java

```diff
@@ -547,12 +543,13 @@ private void processReplicaMetadataResponse(Set<StoreKey> missingRemoteStoreKeys
       } else if (!missingRemoteStoreKeys.contains(messageInfo.getStoreKey())) {
         // the key is present in the local store. Mark it for deletion if it is deleted in the remote store and not
         // deleted yet locally
-        if (messageInfo.isDeleted() && !remoteReplicaInfo.getLocalStore().isKeyDeleted(localKey)) {
+        boolean deletedLocally = remoteReplicaInfo.getLocalStore().isKeyDeleted(localKey);
+        if (messageInfo.isDeleted() && !deletedLocally) {
           MessageFormatInputStream deleteStream =
```

Comment:
New a method `applyDelete()` ?

Reply:
This class probably needs a refactor (for e.g., it uses blocking I/O). I don't want to modify anything that already exists in the context of this change.